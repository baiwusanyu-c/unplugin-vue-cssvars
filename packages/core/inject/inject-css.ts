import hash from 'hash-sum'
import { transformInjectCSS } from '../transform/transform-inject-css'
import {parseCssVars, parseImports} from '../parser'
import type { MagicStringBase } from 'magic-string-ast'
import type { TInjectCSSContent } from '../runtime/process-css'
import type { SFCDescriptor } from '@vue/compiler-sfc'
import type { TMatchVariable } from '../parser'
declare type PCSSVARARR = Array<{
  start: number,
  end: number,
  offset: number,
  variable: string}>
export function injectCSSOnServer(
  mgcStr: MagicStringBase,
  vbindVariableList: TMatchVariable | undefined,
  isHMR: boolean,
) {

  const pCssVarsArr:PCSSVARARR = []
  parseCssVars([mgcStr.toString()], {
    getIndex(start: number, end: number, offset, variable) {
      pCssVarsArr.push({start, end, offset, variable})
    }
  })

  pCssVarsArr.forEach(pca => {
    if(vbindVariableList){
      for (let i = 0; i < vbindVariableList.length; i++) {
        const vbVar = vbindVariableList[i]
        // 样式文件修改后，style热更新可能会先于 sfc 热更新运行，这里先设置hash
        // 详见 packages/core/index.ts的 handleHotUpdate
        if (!vbVar.hash && isHMR)
          vbVar.hash = hash(vbVar.value + vbVar.has)

        if(vbVar.value === pca.variable){
          const offset = pca.offset
          const start = pca.start + offset
          const end = pca.end + offset
          vbVar.hash && (mgcStr = mgcStr.overwrite(start, end, `--${vbVar.hash}`))
          break
        }
      }
    }
  })

  mgcStr = mgcStr.replace(/v-bind-m\s*\(/g, "var(");
  return mgcStr
}

export function injectCssOnBuild(
  mgcStr: MagicStringBase,
  injectCSSContent: TInjectCSSContent | null,
  descriptor: SFCDescriptor | null) {
  if (!injectCSSContent && !descriptor) return mgcStr
  const cssContent = [...injectCSSContent!]
  let resCode = ''

  descriptor!.styles && descriptor!.styles.forEach((value, index) => {
    let injectCssCode = ''
    cssContent.forEach((value) => {
      if (value.styleTagIndex === index)
        injectCssCode = `${injectCssCode}\n${transformInjectCSS(value.content, parseImports(value.content).imports)}`
    })
    const lang = value.lang || 'css'
    const scoped = value.scoped ? 'scoped' : ''
    resCode = `${resCode}\n<style lang="${lang}" ${scoped}> ${injectCssCode}\n${transformInjectCSS(value.content, parseImports(value.content).imports)} </style>`
  })
  resCode && (mgcStr = removeStyleTagsAndContent(mgcStr))
  return mgcStr.prependRight(mgcStr.length(), resCode)
}

export function removeStyleTagsAndContent(mgcStr: MagicStringBase): MagicStringBase {
  return mgcStr.replace(/<style\b[^>]*>[\s\S]*?<\/style>/gi, '')
    .replace(/<style\b[^>]*>/gi, '').replace(/<\/style>/gi, '')
}
