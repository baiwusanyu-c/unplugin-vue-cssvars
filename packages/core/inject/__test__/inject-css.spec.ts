import { describe, expect, test } from 'vitest'
import { injectCssOnBuild, injectCssOnServer, removeStyleTagsAndContent } from '../inject-css'
describe('inject-css', () => {
  test('injectCssOnServer: basic', () => {
    const code = 'v-bind-m(foo)'
    const vbindVariableList = [{ value: 'foo', hash: 'hash' }]
    expect(injectCssOnServer(code, vbindVariableList as any)).toBe('var(--hash)')
  })

  test('injectCssOnServer: vbindVariableList is undefined', () => {
    const code = 'v-bind-m(foo)'
    expect(injectCssOnServer(code, undefined)).toBe(code)
  })

  test('removeStyleTagsAndContent: basic', () => {
    const html = `
      <html>
        <head>
          <title>Test</title>
          <style>
            body {
              background-color: red;
            }
          </style>
        </head>
        <body>
          <h1>Hello, world!</h1>
          <p>This is a test.</p>
          <style type="text/css">
            p {
              font-size: 16px;
            }
          </style>
        </body>
      </html>
    `

    const expectedHtml = `
      <html>
        <head>
          <title>Test</title>
          
        </head>
        <body>
          <h1>Hello, world!</h1>
          <p>This is a test.</p>
          
        </body>
      </html>
    `

    expect(removeStyleTagsAndContent(html)).toEqual(expectedHtml)
  })

  test('removeStyleTagsAndContent: empty HTML', () => {
    expect(removeStyleTagsAndContent('')).toEqual('')
  })

  test('removeStyleTagsAndContent: not modify HTML without style tags', () => {
    const html = `
      <html>
        <head>
          <title>Test</title>
        </head>
        <body>
          <h1>Hello, world!</h1>
          <p>This is a test.</p>
        </body>
      </html>
    `

    expect(removeStyleTagsAndContent(html)).toEqual(html)
  })

  test('injectCssOnBuild: basic', () => {
    const code = '<style lang="scss">\n'
      + '/* foo.scss -> test2.css -> test.css */\n'
      + '/* foo.scss -> test.scss -> test2.css */\n'
      + '\n'
      + '/*@import "./assets/less/less-foo";*/\n'
      + 'div {\n'
      + '  color: v-bind(color)\n'
      + '}\n'
      + "@import './assets/scss/foo.scss';\n"
      + '</style>'
    const injectCSSContent = new Set([{ content: '@import \'./assets/scss/foo.scss\';\nbody { background-color: black; }', lang: 'scss' }])
    const descriptor = {
      styles: [
        {
          lang: 'scss',
          content: '/* foo.scss -> test2.css -> test.css */\n'
            + '/* foo.scss -> test.scss -> test2.css */\n'
            + '\n'
            + '/*@import "./assets/less/less-foo";*/\n'
            + 'div {\n'
            + '  color: v-bind(color)\n'
            + '}\n'
            + "@import './assets/scss/foo.scss';",
        },
      ],
    }
    const result = injectCssOnBuild(code, injectCSSContent, descriptor as any)
    expect(result).toMatchSnapshot()
  })

  test('injectCssOnBuild: no styles', () => {
    const code = 'test'
    const injectCSSContent = new Set([{ content: '@import \'./assets/scss/foo.scss\';\nbody { background-color: black; }', lang: 'scss' }])
    const descriptor = {
      styles: null,
    }
    const result = injectCssOnBuild(code, injectCSSContent, descriptor as any)
    expect(result).toMatchSnapshot()
  })

  test('injectCssOnBuild: no lang', () => {
    const code = '<style lang="scss">\n'
      + '/* foo.scss -> test2.css -> test.css */\n'
      + '/* foo.scss -> test.scss -> test2.css */\n'
      + '\n'
      + '/*@import "./assets/less/less-foo";*/\n'
      + 'div {\n'
      + '  color: v-bind(color)\n'
      + '}\n'
      + "@import './assets/scss/foo.scss';\n"
      + '</style>'
    const injectCSSContent = new Set([{ content: '@import \'./assets/scss/foo.scss\';\nbody { background-color: black; }', lang: 'scss' }])
    const descriptor = {
      styles: [
        {
          lang: null,
          content: '/* foo.scss -> test2.css -> test.css */\n'
            + '/* foo.scss -> test.scss -> test2.css */\n'
            + '\n'
            + '/*@import "./assets/less/less-foo";*/\n'
            + 'div {\n'
            + '  color: v-bind(color)\n'
            + '}\n'
            + "@import './assets/scss/foo.scss';",
        },
      ],
    }
    const result = injectCssOnBuild(code, injectCSSContent, descriptor as any)
    expect(result).toMatchSnapshot()
  })
})
