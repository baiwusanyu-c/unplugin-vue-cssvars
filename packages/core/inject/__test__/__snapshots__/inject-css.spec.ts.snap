// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inject-css > injectCssOnBuild: basic 1`] = `
"

<style lang=\\"scss\\">/* foo.scss -> test2.css -> test.css */
/* foo.scss -> test.scss -> test2.css */

/*@import \\"./assets/less/less-foo\\";*/
div {
  color: v-bind(color)
}
</style>
<style lang=\\"scss\\">
body { background-color: black; }</style>"
`;

exports[`inject-css > injectCssOnBuild: no lang 1`] = `
"

<style lang=\\"css\\">/* foo.scss -> test2.css -> test.css */
/* foo.scss -> test.scss -> test2.css */

/*@import \\"./assets/less/less-foo\\";*/
div {
  color: v-bind(color)
}
</style>
<style lang=\\"scss\\">
body { background-color: black; }</style>"
`;

exports[`inject-css > injectCssOnBuild: no styles 1`] = `
"test

<style lang=\\"scss\\">
body { background-color: black; }</style>"
`;
